Telemetry state change event

aaronsteers opened the pull request:
Resolves https://github.com/meltano/internal-general/issues/338

This adds a one-time 'telemetry_state_change' event when the value of either project_id or send_anonymous_usage_stats changes.

~~Note: this borrows from #5941 and targets that branch.~~

~~Plan is that #5941 will merge first and then this will be updated to target `main`.~~ Done.

TODO:

- [x] Feature development
- [x] Define Snowplow schema for the new event type
- [ ] Add tests
- [x] Manual sanity check
- [x] Register the new event type with Snowcatcloud

edgarrmondragon committed:
Add general context schema

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

edgarrmondragon committed:
Create general Snowplow context for Meltano events

WillDaSilva committed:
Merge branch 'drop-ga-tracker' of github.com:meltano/meltano into drop-ga-tracker

WillDaSilva committed:
Sort dependencies & add `backports-cached-property` and `tzlocal`

WillDaSilva committed:
Add return type hints for path-obtaining methods

WillDaSilva committed:
Add initial Snowplow schemas for command events, the environment context, and the plugin context

WillDaSilva committed:
Add `EnvironmentContext` class

This class can be instantiated with a project to fetch information about
the project and its environment.

WillDaSilva committed:
Add `MeltanoTracker` class

It will replace the `GoogleAnalyticsTracker`, and add at least a little
bit of abstraction between the places in the codebase using the tracker,
and the backend being used for tracking, which is currently Snowplow.

WillDaSilva committed:
Move `hash_sha256` into `meltano.core.utils`

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

WillDaSilva committed:
Add `MeltanoTracker.track_command_event`

WillDaSilva committed:
Move `MeltanoTracker.with_contexts`

WillDaSilva committed:
Separate project context from environment context

WillDaSilva committed:
Turn the Snowplow command event into a context

WillDaSilva committed:
Add schema for slim command events

WillDaSilva committed:
Disallow additional properties not specified by the environment or plugins context schemas

WillDaSilva committed:
Remove elements from environment context

WillDaSilva committed:
`GoogleAnalyticsTracker` -> `LegacyTracker`

WillDaSilva committed:
Correct DockerHub links

WillDaSilva committed:
Add `is_dev_build` and `is_ci_environment` to Snowplow environment context

WillDaSilva committed:
Add TODO comment to remove `MeltanoTracker.track_struct_event` later

WillDaSilva committed:
Add additional admin to telemetry schema readme

WillDaSilva committed:
Switch cached property dependency & update Poetry lock

WillDaSilva committed:
Move tracker out of `__init__.py`

WillDaSilva committed:
Address misc review comments

WillDaSilva committed:
Do not collect platform node

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

Commit:
Merge branch 'main' into drop-ga-tracker

Commit:
Move legacy tracker to its own package

Commit:
flak8 fixes

Commit:
Hash env name in project context

Commit:
Update schema's from PR feedback

Commit:
Restructure on disk layout of schema's to play nice with igluctl

Commit:
Quick test to make sure initializing the tracker doesn't kaboom

Commit:
Maybe fix tests

Commit:
Linter fixes and removal of win specific core discovery

Commit:
Test meltano invoke implementation and more linting

Commit:
Fix 3.7 test failure due to mock import rewrites

Commit:
snowplow-micro support

Commit:
Fires a known good event!

Commit:
Update src/meltano/core/tracking/README.md

Co-authored-by: Aaron ("AJ") Steers <aj@meltano.com>

Commit:
Update src/meltano/core/tracking/iglu-client-embedded/schemas/com.meltano/environment_context/jsonschema/1-0-0

Co-authored-by: Aaron ("AJ") Steers <aj@meltano.com>

Commit:
Merge branch 'main' into drop-ga-tracker

aaronsteers committed:
feat: send telemetry state change event (early draft)

aaronsteers committed:
update from main

aaronsteers committed:
move client id logic to tracker

aaronsteers committed:
first draft

aaronsteers committed:
ready for review

aaronsteers committed:
add json schema file

aaronsteers committed:
revert extra import

aaronsteers committed:
convert to str

aaronsteers committed:
cleanup docstrings

edgarrmondragon commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -44,6 +51,17 @@ def check_url(url: str) -> bool:
     return bool(re.match(URL_REGEX, url))
 
 
+class AnalyticsSettings(NamedTuple):
```
Wdyt about calling this class `Telemetry settings`?

aaronsteers replied:
Sure! Initially it was named after the analytics.json file, but TelemetrySettings seems more appropriate, I think. 👍

edgarrmondragon commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -194,3 +263,103 @@ def track_command_event(self, event_json: dict[str, Any]) -> None:
+        """Return path to the 'analytics.json' file."""
+        return self.project.meltano_dir() / "analytics.json"
+
+    @property
+    def _analytics_json_settings(self) -> AnalyticsSettings:
```
IMO it's a bit weird to have a "private" computed property. What do you think of making this a proper method since it's only called once anyway?

@WillDaSilva or @pandemicsyn might have opinions too 🙂

pandemicsyn replied:
+1 I'd prefer this one be a method, IMHO this does a lot for being a "property" .

aaronsteers replied:
Agreed - this one grew and morphed quite a lot. I tidied it and moved the helper function out to its own private method.

Commit:
Init pass at a PluginsTrackingContext class and sample meltano invoke usage

Commit:
Merge branch 'main' into 3366-add-snowplow-events-at-the-block-level

Commit:
Merge branch 'main' into 3366-add-snowplow-events-at-the-block-level

Commit:
Add some safe guards against funky plugins

Commit:
Cleanup

Commit:
Have run fire block events

pandemicsyn commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -3,10 +3,13 @@
 from contextlib import contextmanager
-from typing import Any
+from pathlib import Path
+from typing import Any, NamedTuple, Optional, Tuple, Union
```
@aaronsteers  that from __future__ import annotations allows new style annotations, and we've been slowly moving the code base to those. That allows referring list , tuple, and friends without having to import anything from the typing package.

If you're running the pre-commit hook pyupgrade will catch a lot of these for you, but figured I'd give a ya heads up.

aaronsteers replied:
Updated per your suggestion. Thanks!

I considered putting them in the new style before, but my IDE linter is complaining about the new types. Going forward, I may need to find a fix for the mypy lint issue, perhaps using a higher Python version for my development, since it seems mypy prefers 3.9 or 3.10.

pandemicsyn commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -44,6 +51,17 @@ def check_url(url: str) -> bool:
+    These are stored within `analytics.json`.
+    """
+
+    client_id: Optional[uuid.UUID]
```
for ref, these can convert to:

```
client_id: uuid.UUID | None
```

pandemicsyn commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -194,3 +263,103 @@ def track_command_event(self, event_json: dict[str, Any]) -> None:
+                    return None
+
+                logger.warn(
+                    f"The value '{from_val}' in 'analytics.json' was of type "
```
can we use a structured log entry here. Something like:

```logger.warn("Unexpected analytics.json setting type", from_val=from_val, current_type=type(from_val))```

pandemicsyn commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -194,3 +263,103 @@ def track_command_event(self, event_json: dict[str, Any]) -> None:
+                if warn:
+                    log_fn = logger.warning
+
+                log_fn(
```
Ditto the structlog entry here would be nice.

edgarrmondragon commented:
The commit history in this one has things it shouldn't, right? Does it need to be rebased?

Commit:
Check-in missing block event

Commit:
Fix up option entries

Commit:
Merge branch 'main' into 3366-add-snowplow-events-at-the-block-level

Commit:
Add hashed namespace to plugins context

Commit:
Simplify using option_keys

Commit:
Add plugins tracking contexts per block

Commit:
Discussion/pr feedback

Commit:
Merge branch 'main' into 3366-add-snowplow-events-at-the-block-level

Commit:
Track failures in invoke

aaronsteers committed:
Merge branch 'main' into telemetry-state-change-event

aaronsteers committed:
cleanup per pr feedback

pandemicsyn commented:
@WillDaSilva heads up on https://github.com/meltano/meltano/pull/6028 which should hopefully land this morning, it'll probably give you some merge conflicts.

WillDaSilva committed:
Merge branch 'main' into telemetry-state-change-event

WillDaSilva committed:
Add maxLength constraints to `telemetry_state_change_event`

Commit:
Add basic test of PluginsContext generation and fix run capture on
exception

WillDaSilva committed:
Address lint issues

Commit:
Distinguish between RunnerErrors and all others

Commit:
Capture BlockSetValidationErrors raised from find

Commit:
Merge branch 'main' into 3366-add-snowplow-events-at-the-block-level

Commit:
Fixup block event schema to include maxchar limit for type

WillDaSilva committed:
Misc changes to telemetry state change event schema

WillDaSilva committed:
Format exceptions with traceback in logs

WillDaSilva committed:
Send telemetry state change event directly without context

We ignore the regular `can_track` check because it would prevent us from
learning that tracking was disabled, and we avoid sending any
Snowplow contexts because those contain more info than we want from this
event.

WillDaSilva committed:
Merge branch '3366-add-snowplow-events-at-the-block-level' into telemetry-state-change-event

Commit:
Update src/meltano/core/tracking/tracker.py

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

Commit:
Update src/meltano/core/utils/__init__.py

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

WillDaSilva committed:
Merge remote-tracking branch 'origin/3366-add-snowplow-events-at-the-block-level' into telemetry-state-change-event

Commit:
Updates to play nice with PR suggested commits

Commit:
Merge branch '3366-add-snowplow-events-at-the-block-level' of github.com:meltano/meltano into 3366-add-snowplow-events-at-the-block-level

WillDaSilva committed:
Merge remote-tracking branch 'origin/3366-add-snowplow-events-at-the-block-level' into telemetry-state-change-event

Commit:
Update src/meltano/core/tracking/tracker.py

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

WillDaSilva committed:
Fix telemetry state change (project ID change) firing on every run

WillDaSilva committed:
Merge remote-tracking branch 'origin/3366-add-snowplow-events-at-the-block-level' into telemetry-state-change-event

WillDaSilva commented:
@aaronsteers @pandemicsyn @pnadolny13 The new event type `telemetry_state_change_event` has been added to Snowcatcloud. Pending your approval, I'll put it into production mode.

pandemicsyn commented:
@WillDaSilva fyi - https://github.com/meltano/meltano/pull/6028 landed so probably have a little merge/commit clean up.

WillDaSilva committed:
Merge branch 'main' into telemetry-state-change-event

WillDaSilva commented:
@aaronsteers I am unable to add you as a reviewer.

aaronsteers commented:
@WillDaSilva - Because I authored it I guess I cannot be an approver. I'll review now and add any comments I see.

aaronsteers commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -100,17 +118,80 @@ def __init__(
 
+        self.telemetry_state_change_check(stored_telemetry_settings)
+
+    @cached_property
+    def send_anonymous_usage_stats(self) -> bool:
+        """Return whether anonymous usages stats are enabled (bool).
+
+        - Return the value from 'send_anonymous_usage_stats', if set.
+        - Otherwise the opposite of 'tracking_disabled', if set.
+        - Otherwise return 'True'
+
+        Returns:
+            True if anonymous usage stats are enabled.
+        """
+        if self.settings_service.get("send_anonymous_usage_stats", None) is not None:
+            return self.settings_service.get("send_anonymous_usage_stats")
+
+        if self.settings_service.get("tracking_disabled", None) is not None:
+            return not self.settings_service.get("tracking_disabled")
+
+        return True
```
Yes, this is the bit that resolves the removal of the `env_alias` `MELTANO_TRACKING_DISABLED`.

aaronsteers reviewed (commented):
Looks good! ✅

pandemicsyn reviewed (approved):
lgtm! Also double checked,  I think i see the correct # of events at the correct times, in snowplow micro
